Command completion race condition

Seeing a rare racecondition likely related to `Observable.zip` in https://github.com/d4rken/RxShell/blob/master/core/src/main/java/eu/darken/rxshell/cmd/CmdProcessor.java#L111

```
1513428431737 V/RXS:RxCmdShell: open()
1513428431738 V/RXS:RxShell: open()
1513428431738 V/RXS:RxProcess: open()
1513428431740 V/RXS:RxProcess: processCreator:onNext(Process[pid=5851, hasExited=false])
1513428431740 D/RXS:RxProcess: open():doOnSuccess RxProcess.Session(process=Process[pid=5851, hasExited=false])
1513428431740 V/RXS:RxProcess:Session: input()
1513428431740 V/RXS:RxProcess:Session: output()
1513428431740 V/RXS:RxProcess:Session: error()
1513428431740 V/RXS:RxShell:Session: errorLines()
1513428431740 V/RXS:RxProcess:Session: destroy()
1513428431740 V/RXS:RxProcess:Session: waitFor()
1513428431740 V/RXS:RxShell:Session: waitFor()
1513428431740 V/RXS:RxShell: open():doOnSuccess RxShell.Session(processSession=RxProcess.Session(process=Process[pid=5851, hasExited=false]))
1513428431740 V/RXS:RxShell:Session: waitFor()
1513428431741 D/RXS:RxShell:Session: writeLine(line=V5B9D812A9C51=/data/user/0/eu.thedarken.sdm/files/toybox_sdm, flush=true)
1513428431741 V/RXS:CmdProcessor: attach(RxShell.Session(processSession=RxProcess.Session(process=Process[pid=5851, hasExited=false])))
1513428431741 V/RXS:RxShell:Session: waitFor()
1513428431741 V/RXS:RxShell:Session: waitFor()
1513428431741 V/RXS:RxShell:Session: cancel()
1513428431741 V/RXS:RxCmdShell: open():doOnSuccess eu.darken.rxshell.cmd.RxCmdShell$Session@9b97c1513428431741 V/RXS:RxCmdShell:Session: waitFor()
1513428431742 D/RXS:RxProcess:Session: Waiting for Process[pid=5851, hasExited=false] to exit.
1513428431742 D/RXS:CmdProcessor: Submitted: Cmd(timeout=0, commands=[$V5B9D812A9C51 test ! -e '/storage/emulated/0' || $V5B9D812A9C51 find -L '/storage/emulated/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N", $V5B9D812A9C51 test ! -e '/data/user_de/0' || $V5B9D812A9C51 find -L '/data/user_de/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N", $V5B9D812A9C51 test ! -e '/data/user/0' || $V5B9D812A9C51 find -L '/data/user/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N"])
1513428431742 D/RXS:CmdProcessor: Processing: Cmd(timeout=0, commands=[$V5B9D812A9C51 test ! -e '/storage/emulated/0' || $V5B9D812A9C51 find -L '/storage/emulated/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N", $V5B9D812A9C51 test ! -e '/data/user_de/0' || $V5B9D812A9C51 find -L '/data/user_de/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N", $V5B9D812A9C51 test ! -e '/data/user/0' || $V5B9D812A9C51 find -L '/data/user/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N"])
1513428431742 V/RXS:RxShell:Session: outputLines()
1513428431742 V/RXS:Harvester:Output: request(9223372036854775807)
1513428431742 V/RXS:RxShell:Session: errorLines()
1513428431742 V/RXS:Harvester:Error: request(9223372036854775807)
1513428431742 D/RXS:RxShell:Session: writeLine(line=$V5B9D812A9C51 test ! -e '/storage/emulated/0' || $V5B9D812A9C51 find -L '/storage/emulated/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N", flush=false)
1513428431742 D/RXS:RxShell:Session: writeLine(line=$V5B9D812A9C51 test ! -e '/data/user_de/0' || $V5B9D812A9C51 find -L '/data/user_de/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N", flush=false)
1513428431742 D/RXS:RxShell:Session: writeLine(line=$V5B9D812A9C51 test ! -e '/data/user/0' || $V5B9D812A9C51 find -L '/data/user/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N", flush=false)
1513428431742 D/RXS:RxShell:Session: writeLine(line=echo fb6e4bdf-16ae-4f0d-b1b8-c3445f5d7ec1 $?, flush=false)
1513428431742 D/RXS:RxShell:Session: writeLine(line=echo fb6e4bdf-16ae-4f0d-b1b8-c3445f5d7ec1 >&2, flush=true)
1513428431775 V/RXS:Harvester:Error: Permission denied
1513428431775 D/RXS:Harvester:Error: Permission denied
1513428431776 V/RXS:RxShell: LineStream:error onComplete()
1513428431776 V/RXS:RxShell: LineStream:output onComplete()
1513428431776 V/RXS:Harvester:Error: onComplete()
1513428431776 V/RXS:Harvester:Output: onComplete()
1513428431776 V/RXS:Harvester:Error: onError(java.io.IOException: Upstream completed prematurely.)
1513428431776 V/RXS:Harvester:Output: onError(java.io.IOException: Upstream completed prematurely.)
1513428431776 V/RXS:CmdProcessor: errorLines():doOnEach: OnNextNotification[eu.darken.rxshell.cmd.Harvester$Crop@519a13a]
1513428431776 V/RXS:CmdProcessor: outputLine():doOnEach: OnNextNotification[eu.darken.rxshell.cmd.OutputHarvester$Crop@7e581eb]
1513428431776 V/RXS:RxProcess: processCreator:onComplete()
1513428431776 V/RXS:CmdProcessor: errorLines():doOnEach: OnCompleteNotification
1513428431776 V/RXS:Harvester:Error: cancel()
1513428431776 V/RXS:RxProcess: cancel()
1513428431777 D/RXS:RootKiller: kill(Process[pid=5851 ,hasExited=true, exitcode=1])
1513428431777 D/RXS:RootKiller: Process is no longer alive, skipping kill.
1513428431777 V/RXS:RxProcess: Process finished, clearing session
1513428431777 V/RXS:RxShell: LineStream:error onCancel()
1513428431777 D/RXS:RxProcess:Session: Exitcode: 1, Process: Process[pid=5851 ,hasExited=true, exitcode=1]
1513428431777 V/RXS:RxProcess:Session: waitFor():doOnSuccess 1513428431777 V/RXS:RxShell:Session: waitFor():doOnSuccess 1513428431777 V/RXS:CmdProcessor: Attached session ended!
1513428431777 V/RXS:CmdProcessor: Poison pill!
1513428431777 D/RXS:CmdProcessor: Post zip: OnNextNotification[QueueCmd(command=Cmd(timeout=0, commands=[$V5B9D812A9C51 test ! -e '/storage/emulated/0' || $V5B9D812A9C51 find -L '/storage/emulated/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N", $V5B9D812A9C51 test ! -e '/data/user_de/0' || $V5B9D812A9C51 find -L '/data/user_de/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N", $V5B9D812A9C51 test ! -e '/data/user/0' || $V5B9D812A9C51 find -L '/data/user/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N"]), exitCode=-3, output.size()=null, errors.size()=null)]
1513428431777 V/RXS:CmdProcessor: onNext(QueueCmd(command=Cmd(timeout=0, commands=[$V5B9D812A9C51 test ! -e '/storage/emulated/0' || $V5B9D812A9C51 find -L '/storage/emulated/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N", $V5B9D812A9C51 test ! -e '/data/user_de/0' || $V5B9D812A9C51 find -L '/data/user_de/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N", $V5B9D812A9C51 test ! -e '/data/user/0' || $V5B9D812A9C51 find -L '/data/user/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N"]), exitCode=-3, output.size()=null, errors.size()=null))
1513428431777 V/RXS:RxShell:Session: cancel()
1513428431777 I/RXS:CmdProcessor: Processed: Cmd.Result(cmd=Cmd(timeout=0, commands=[$V5B9D812A9C51 test ! -e '/storage/emulated/0' || $V5B9D812A9C51 find -L '/storage/emulated/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N", $V5B9D812A9C51 test ! -e '/data/user_de/0' || $V5B9D812A9C51 find -L '/data/user_de/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N", $V5B9D812A9C51 test ! -e '/data/user/0' || $V5B9D812A9C51 find -L '/data/user/0' -mindepth 1 -print0 | $V5B9D812A9C51 xargs -0 -r -s 65536 $V5B9D812A9C51 stat -c "%a:%F:%d:%i:%u:%g:%s:%B:%b:%Y:%N"]), exitcode=-3, output.size()=null, errors.size()=null)
1513428431777 V/RXS:RxProcess:Session: destroy():doOnComplete
1513428431777 D/RXS:CmdProcessor: Post zip: OnCompleteNotification
1513428431777 V/RXS:RxShell:Session: cancel():doOnComplete
1513428431777 V/RXS:CmdProcessor: onComplete()
1513428431777 V/RXS:CmdProcessor: outputLine():doOnEach: OnCompleteNotification
1513428431777 V/RXS:Harvester:Output: cancel()
io.reactivex.exceptions.OnErrorNotImplementedException: Upstream completed prematurely.
	at io.reactivex.e.b.a$af.accept(Functions.java:1704)
	at io.reactivex.e.h.k.onError(LambdaSubscriber.java:79)
	at io.reactivex.i.b.onError(PublishProcessor.java:2325)
	at eu.darken.rxshell.cmd.Harvester$BaseSub.onError(Harvester.java:96)
	at eu.darken.rxshell.cmd.Harvester$BaseSub.onComplete(Harvester.java:102)
	at io.reactivex.e.j.k.a(HalfSerializer.java:91)
	at io.reactivex.e.h.s.onComplete(StrictSubscriber.java:109)
	at io.reactivex.e.e.b.co$a.onComplete(FlowableRefCount.java:82)
	at io.reactivex.e.e.b.cg$c.a(FlowablePublish.java:366)
	at io.reactivex.e.e.b.cg$c.c(FlowablePublish.java:429)
	at io.reactivex.e.e.b.cg$c.onComplete(FlowablePublish.java:252)
	at io.reactivex.e.e.b.dm$a.onComplete(FlowableSubscribeOn.java:108)
	at io.reactivex.e.e.b.aa$a.c(FlowableCreate.java:262)
	at io.reactivex.e.e.b.aa$a.n_(FlowableCreate.java:254)
	at eu.darken.rxshell.shell.RxShell.lambda$makeLineStream$6$RxShell(RxShell.java:190)
	at eu.darken.rxshell.shell.RxShell$$Lambda$6.subscribe(Unknown Source:4)
	at io.reactivex.e.e.b.aa.subscribeActual(FlowableCreate.java:72)
	at io.reactivex.i.subscribe(Flowable.java:13041)
	at io.reactivex.i.subscribe(Flowable.java:12987)
	at io.reactivex.e.e.b.dm$a.run(FlowableSubscribeOn.java:82)
	at io.reactivex.e.g.l.run(ScheduledRunnable.java:61)
	at io.reactivex.e.g.l.call(ScheduledRunnable.java:52)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
	at java.lang.Thread.run(Thread.java:764)
Caused by: java.io.IOException: Upstream completed prematurely.
	... 23 more
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -16,6 +16,10 @@ android:
       - 'android-sdk-license-.+'
       - 'google-gdk-license-.+'
 
+before_install:
+  - chmod +x gradlew
+  - yes | sdkmanager "platforms;android-27"
+
 script: ./gradlew :core:clean :core:assemble :core:test
 
 after_success: ./gradlew :core:jacocoTestReport :core:coveralls

--- a/core/src/main/java/eu/darken/rxshell/cmd/CmdProcessor.java
+++ b/core/src/main/java/eu/darken/rxshell/cmd/CmdProcessor.java
@@ -95,7 +95,7 @@ public class CmdProcessor {
 
                     final Observable<Harvester.Crop> errors = session.errorLines()
                             .compose(upstream -> factory.forError(upstream, item.cmd))
-                            .onErrorReturnItem(new Harvester.Crop(null))
+                            .onErrorReturnItem(new ErrorHarvester.Crop(null))
                             .doOnEach(n -> { if (RXSDebug.isDebug()) Timber.tag(TAG).v("errorLines():doOnEach: %s", n); })
                             .toObservable().cache();
                     errors.subscribe(s -> {}, e -> {});
@@ -108,16 +108,29 @@ public class CmdProcessor {
                         return Observable.just(item.exitCode(Cmd.ExitCode.SHELL_DIED));
                     }
 
-                    Observable<QueueCmd> harvestZipper = Observable.zip(outputs, errors, (out, err) -> item.outputHarvest(out).errorHarvest(err));
+                    Observable<QueueCmd> cropWait = Observable.merge(outputs, errors)
+                            .toList().toObservable()
+                            .map(crops -> {
+                                for (Harvester.Crop crop : crops) {
+                                    if (crop instanceof OutputHarvester.Crop) {
+                                        item.exitCode(((OutputHarvester.Crop) crop).exitCode);
+                                        item.output(crop.buffer);
+                                    } else {
+                                        item.errors(crop.buffer);
+                                    }
+                                }
+                                if (crops.size() != 2) item.exitCode(Cmd.ExitCode.SHELL_DIED);
+                                return item;
+                            });
                     if (item.cmd.getTimeout() > 0) {
-                        harvestZipper = harvestZipper.timeout(item.cmd.getTimeout(), TimeUnit.MILLISECONDS).onErrorReturn(error -> {
+                        cropWait = cropWait.timeout(item.cmd.getTimeout(), TimeUnit.MILLISECONDS).onErrorReturn(error -> {
                             if (error instanceof TimeoutException) {
                                 if (RXSDebug.isDebug()) Timber.tag(TAG).w("Command timed out: %s", item);
                                 return item.exitCode(Cmd.ExitCode.TIMEOUT);
                             } else throw new RuntimeException(error);
                         });
                     }
-                    return harvestZipper;
+                    return cropWait;
                 })
                 .doOnEach(n -> { if (RXSDebug.isDebug()) Timber.tag(TAG).d("Post zip: %s", n); })
                 .subscribe(new Observer<QueueCmd>() {
@@ -174,16 +187,15 @@ public class CmdProcessor {
             return this;
         }
 
-        public QueueCmd outputHarvest(OutputHarvester.Crop out) {
+        public QueueCmd output(List<String> output) {
             if (this.output != null) throw new IllegalStateException("Output already set!");
-            exitCode(out.exitCode);
-            this.output = out.buffer;
+            this.output = output;
             return this;
         }
 
-        public QueueCmd errorHarvest(Harvester.Crop err) {
+        public QueueCmd errors(List<String> errors) {
             if (this.errors != null) throw new IllegalStateException("Errors already set!");
-            this.errors = err.buffer;
+            this.errors = errors;
             return this;
         }
 

